### PR TITLE
M 3714893369 corpus onboarding

### PR DESCRIPTION
**Pull request by thiago-aixplain**

### CODE REVIEW

**Review by krishnadurai — COMMENTED** [review]
@thiago-aixplain here's my first pass in the review. I'll be taking another pass to understand the data, file, dataset and corpus flow. Some of the high level comments are:

1. The asset factory implementation seems incomplete. The present inheritors of this class do not seem to implement the get function.
2. Let us add docstrings to the new Classes and their functions.

Keep in mind that we are in the process of changing the package name from aiXtend to aiXplain.

File: aixtend/__init__.py [hunk 1]
```diff
@@ -19,12 +19,9 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 """
-from .modules.pipeline import Pipeline
-from .modules.model import Model
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/__init__.py:23]: Due to circular import exceptions I was facing. Then I decided to split the paths in two general ones, factories and modules:

from aixplain.modules import Model, Pipeline, Corpus
from aixplain.factoris import ModelFactory, PipelineFactory...

File: aixtend/modules/corpus.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+from aixtend.enums.license import License
+from aixtend.enums.onboard_status import OnboardStatus
+from aixtend.enums.privacy import Privacy
+from aixtend.modules.asset import Asset
+from aixtend.modules.data import Data
+from aixtend.utils.file_utils import _request_with_retry, save_file
+from typing import Any, List, Optional, Text
+
+
+class Corpus(Asset):
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/modules/corpus.py:33]: Done. In the user_doc.md is also explained

File: aixtend/modules/pipeline.py [hunk 3]
```diff
@@ -24,29 +24,42 @@
...
-from typing import Union
-
-
-class Pipeline:
-    def __init__(self, id: str, name:str, api_key: str, url: str = config.PIPELINES_RUN_URL, **additional_info) -> None:
+from aixtend.utils.file_utils import _request_with_retry, path2link
+from typing import Dict, Optional, Text, Union
+
+
+class Pipeline(Asset):
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/modules/pipeline.py:33]: It is already there.

File: aixtend/factories/corpus_factory.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+from aixtend.enums.license import License
+from aixtend.enums.privacy import Privacy
+from aixtend.utils.file_utils import _request_with_retry
+from aixtend.utils import config
+from pathlib import Path
+from tqdm import tqdm
+from typing import Any, Dict, List, Optional, Text, Union
+
+
+class CorpusFactory(AssetFactory):
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/factories/corpus_factory.py:46]: This method is implemented as `create_from_asset_id` I want to change this name to `get` in the future and deprecate the former one.

File: aixtend/factories/dataset_factory.py [hunk 5]
```diff
@@ -42,53 +49,54 @@ def _create_dataset_from_response(cls, response: dict) -> Dataset:
...
+            task (Text): Task of listed datasets
+            input_language (Text, optional): Input language of listed datasets. Defaults to None.
+            output_language (Text, optional): Output language of listed datasets. Defaults to None.
 
         Returns:
             List[Dataset]: List of datasets based on given filters
         """
         try:
-            url = f"{cls.backend_url}/sdk/datasets?pageNumber={page_number}&function={task}"
+            url = os.path.join(cls.backend_url, f"sdk/datasets?pageNumber={page_number}&function={task}")
```
Comment (krishnadurai) [inline aixtend/factories/dataset_factory.py:94]: Please replace this with urljoin

File: aixtend/factories/dataset_factory.py [hunk 6]
```diff
@@ -42,53 +49,54 @@ def _create_dataset_from_response(cls, response: dict) -> Dataset:
         Returns:
             Dataset: Coverted 'Dataset' object
         """
-        return Dataset(response['id'], response['name'], response['description'])
+        return Dataset(
+            id=response["id"],
+            name=response["name"],
+            description=response["description"],
+            function=Function.SPEECH_RECOGNITION,
```
Comment (krishnadurai) [inline aixtend/factories/dataset_factory.py:56]: Is this constant value correct?
```Suggestion
            function=response["function"],
```

File: docs/user/user_doc.md [hunk 7]
```diff
@@ -92,8 +93,24 @@ poll_url = start_response["url"]
 poll_response = pipeline.poll(poll_url)
 ```
 
+## Corpus
+aiXplain has an extensive collection of general-purpose corpora to be explored, processed and used to create task-specific datasets.
+
+The aiXtend SDK allows searching for existing corpora that match a specific criteria. `CorpusFactory` can search for corpora that contain data from a particular language pair or data type.
+
+```python
+from auxtend.enums import DataType
```
Comment (krishnadurai) [inline docs/user/user_doc.md:102]: nit: typo for aixtend

File: aixtend/utils/file_utils.py [hunk 8]
```diff
@@ -14,50 +14,167 @@
...
+        url = config.TEMPFILE_UPLOAD_URL
+
+        headers = {"Authorization": "token " + team_key}
+
+        payload = {"contentType": content_type, "originalName": file_name}
+        r = _request_with_retry("post", url, headers=headers, data=payload)
+        response = r.json()
+        path = response["key"]
+        # Upload data
+        presigned_url = response["uploadUrl"]
```
Comment (krishnadurai) [inline aixtend/utils/file_utils.py:132]: Can you please explain in a comment within the code how you fetch this presigned_url? The logic of this TEMPFILE_UPLOAD_URL isn't too clear from reading the implementation.

File: aixtend/utils/file_utils.py [hunk 9]
```diff
@@ -14,50 +14,167 @@
...
+        payload = open(file_name, "rb").read()
+        r = _request_with_retry("put", presigned_url, headers=headers, data=payload)
+
+        if r.status_code != 200:
+            if nattempts == 0:
+                return upload_data(file_name, content_type, content_encoding, nattempts + 1)
+            else:
+                raise Exception("File Uploading Error: Failure on Uploading to S3.")
+        bucket_name = re.findall(r"https://(.*?).s3.amazonaws.com", presigned_url)[0]
+        s3_link = f"s3://{bucket_name}/{path}"
```
Comment (krishnadurai) [inline aixtend/utils/file_utils.py:145]: Are we allowing external users to directly upload to our S3 buckets (with the help of pre-signed urls)?

File: aixtend/utils/file_utils.py [hunk 10]
```diff
@@ -14,50 +14,167 @@
...
+        return s3_link
+    except Exception as e:
+        if nattempts == 0:
+            return upload_data(file_name, content_type, content_encoding, nattempts + 1)
```
Comment (krishnadurai) [inline aixtend/utils/file_utils.py:149]: Why not write the condition as nattempts > 1 and set nattempts = 2 in the initialization. Would be clearer to the user of this function.

File: aixtend/processes/data_onboarding/onboard_functions.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+    url = os.path.join(config.BACKEND_URL, "sdk/inventory/corpus/onboard")
+
+    r = _request_with_retry("post", url, headers=headers, json=payload)
+    if 200 <= r.status_code < 300:
+        response = r.json()
+
+        corpus_id = response["id"]
+        status = response["status"]
+        # coreengine_payload = response["coreEnginePayload"]
```
Comment (krishnadurai) [inline aixtend/processes/data_onboarding/onboard_functions.py:139]: Can we remove the commented lines?

File: aixtend/processes/data_onboarding/process_audio_files.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    return folder_path + ".tgz"
+
+
+def run(metadata: MetaData, paths: List, folder: Path, batch_size: int = 100) -> Tuple[List[File], int, int, int]:
+    """Process a list of local audio files, compress and upload them to pre-signed URLs in S3
+
+    Args:
+        metadata (MetaData): meta data of the asset
+        paths (List): list of paths to local files
+        folder (Path): local folder to save compressed files before upload them to s3.
```
Comment (krishnadurai) [inline aixtend/processes/data_onboarding/process_audio_files.py:32]: What is batch size used for? How is the data stored in batches?
A supporting document or comment can help implementors understand the code here.

**Review by krishnadurai — CHANGES_REQUESTED** [review]
@thiago-aixplain I appreciate the work you have done with the documentation, especially with the notebook!
Here are a few thoughts on the implementation:
1. When we upload files, we check if the size is less than 10 MB on the client's side. Correct me if I'm wrong, but it doesn't seem that we have appropriate checks on the server side for it. Someone on the client could disable the check and upload larger files.
2. Can you add more in-line code documentation for how data is created in batches for upload? The logic seems hard to follow. A document detailing how the data is physically stored might help readers of this code.
3. In several places we follow the pattern of catching exceptions and re-raising them. Usually, it would be better practice to not catch the exceptions as it helps for better tracebacks.

File: aixtend/factories/corpus_factory.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        schema: List[Union[Dict, MetaData]],
+        ref_data: Optional[List[Any]] = [],
+        tags: Optional[List[Text]] = [],
+        functions: Optional[List[Function]] = [],
+        privacy: Optional[Privacy] = Privacy.PRIVATE,
+    ) -> Dict:
+        """Asynchronous call to Upload a corpus to the user's dashboard.
+
+        Args:
+            name (Text): corpus name
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/factories/corpus_factory.py]: I asked @hadi-aix and Ibrahim about it

File: aixtend/factories/corpus_factory.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+                try:
+                    schema = [MetaData(**metadata) for metadata in schema]
+                except Exception as e:
+                    message = "Data Asset Onboarding Error: Make sure the elements of your schema follows the MetaData class."
+                    logging.error(message)
+                    raise Exception(message)
+
+            if len(ref_data) > 0:
+                if isinstance(ref_data[0], Data):
+                    ref_data = [w.id for w in ref_data]
```
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/factories/corpus_factory.py]: validation will happen on the backend for this case.

File: aixtend/factories/dataset_factory.py (hunk 6 above)
Comment (thiago-aixplain) [reply aixtend/factories/dataset_factory.py]: I do not have the service to read and onboard datasets from the new backend. Once I have it, I will update it.

File: aixtend/utils/file_utils.py (hunk 9 above)
Comment (thiago-aixplain) [reply aixtend/utils/file_utils.py]: Yes. This is a service that Anas implemented. The script calls a backend service which returns me an URL to save my file. This URL with the file keeps available for 14 days before being deleted automatically.

File: aixtend/factories/corpus_factory.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thiago-aixplain) [reply aixtend/factories/corpus_factory.py]: Done

File: docs/docstring/aixtend.factories.md [hunk 15]
```diff
@@ -0,0 +1,540 @@
+# aixtend.factories package
```
Comment (krishnadurai) [inline docs/docstring/aixtend.factories.md:1]: You need to regenerate the docstring files.

File: aixplain/processes/data_onboarding/process_text_files.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+    Args:
+        metadata (MetaData): meta data of the asset
+        paths (List): list of paths to local files
+        folder (Path): local folder to save compressed files before upload them to s3.
+
+    Returns:
+        Tuple[List[File], int]: list of s3 links and data colum index
+    """
+    # logging.info(f"Data Asset Onboarding: Processing \"{metadata.name}\".")
```
Comment (krishnadurai) [inline aixplain/processes/data_onboarding/process_text_files.py:49]: Remove commented code or make it a debug logging statement.

File: aixplain/processes/data_onboarding/process_audio_files.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    return folder_path + ".tgz"
+
+
+def run(metadata: MetaData, paths: List, folder: Path, batch_size: int = 100) -> Tuple[List[File], int, int, int]:
+    """Process a list of local audio files, compress and upload them to pre-signed URLs in S3
+
+    Explanation:
+        Each audio on "paths" is processed. If the audio is in a public link, this link is added into an index CSV file.
+        If the audio is in a local path, it will be copied into a local folder and its path will be added to the index CSV file.
+        The audios are processed in batches such that at each "batch_size" audios, the index CSV file is uploaded into a pre-signed URL in s3 and reset.
```
Comment (krishnadurai) [inline aixplain/processes/data_onboarding/process_audio_files.py:32]: Cut-paste an example that already exists for this kind of batched storage in the developer documentation.

File: aixplain/factories/pipeline_factory.py [hunk 18]
```diff
@@ -77,7 +82,16 @@ def create_asset_from_id(cls, pipeline_id: str) -> Pipeline:
                 message = "Pipeline Creation: Unspecified Error"
             logging.error(message)
             raise Exception(f"Status {status_code}: {message}")
-        
+
+    @classmethod
+    def create_asset_from_id(cls, pipeline_id: Text) -> Pipeline:
```
Comment (krishnadurai) [inline aixplain/factories/pipeline_factory.py:87]: +1

**Review by krishnadurai — COMMENTED** [review]
@thiago-aixplain we are almost good to go, bar the few comments that you can address quickly in the review. Additionally, did we look into this issue?

> When we upload files, we check if the size is less than 10 MB on the client's side. Correct me if I'm wrong, but it doesn't seem that we have appropriate checks on the server side for it. Someone on the client could disable the check and upload larger files.

**Review by krishnadurai — APPROVED** [review]
